efi/image.go

```diff
@@ -0,0 +1,240 @@
+	return out
+}
+
+type snapModelParams []secboot.SnapModel
+
+// SnapModelParams returns a ImageLoadParams for the specifed snap models.
+func SnapModelParams(models ...secboot.SnapModel) ImageLoadParams {
+	return snapModelParams(models)
+}
+
+func (p snapModelParams) apply(params ...loadParams) []loadParams {
+	var out []loadParams
+	for _, model := range []secboot.SnapModel(p) {
+		p := make([]loadParams, len(params))
+		copy(p, params)
+		for i := range p {
+			p[i].SnapModel = model
+		}
+		out = append(out, p...)
+	}
+	return out
+}
+
+// ImageLoadEventSource corresponds to the source of a ImageLoadEvent.
+// XXX: This exists temporarily until efi/secureboot_policy.go has been ported to the new
+// profile generation implementation.
+type ImageLoadEventSource int
+
+const (
+	// Firmware indicates that the source of a ImageLoadEvent was platform firmware, via the EFI_BOOT_SERVICES.LoadImage()
+	// and EFI_BOOT_SERVICES.StartImage() functions, with the subsequently executed image being verified against the signatures
+	// in the EFI authorized signature database.
+	Firmware ImageLoadEventSource = iota
+
+	// Shim indicates that the source of a ImageLoadEvent was shim, without relying on EFI boot services for loading, verifying
+	// and executing the subsequently executed image. The image is verified by shim against the signatures in the EFI authorized
+	// signature database, the MOK database or shim's built-in vendor certificate before being executed directly.
+	Shim
+)
+
+func (s ImageLoadEventSource) apply(params ...loadParams) []loadParams {
+	var out []loadParams
+	for _, param := range params {
+		param.Source = s
+		out = append(out, param)
+	}
+	return out
+}
+
+// ImageLoadEvent corresponds to the execution of an image during the boot
+// process, before ExitBootServices.
+type ImageLoadEvent interface {
+	// Next specifies a set of images that are permitted to be executed by the
+	// image associated with this event.
+	Next(images ...ImageLoadEvent) ImageLoadEvent
+
+	source() Image
+	next() []ImageLoadEvent
+	params() imageLoadParamsSet
+}
+
+// NewImageLoadEvent returns a new ImageLoadEvent for the specified image that will
+// be executed during the boot process, before ExitBootServices. The caller can specify
+// optional parameters that will apply to the specified image and which will be inherited
+// by subsequent images (added by [ImageLoadEvent.Next]. The supplied parameters will
+// override any existing ones that would be inherited by this image. Parameters that
+// provide multiple values will automatically create branches in the profile.
+func NewImageLoadEvent(image Image, params ...ImageLoadParams) ImageLoadEvent {
+	return &baseImageLoadEvent{
+		sourceImage: image,
+		loadParams:  params}
+}
+
+type baseImageLoadEvent struct {
+	sourceImage Image
+	nextImages  []ImageLoadEvent
+	loadParams  imageLoadParamsSet
+}
+
+func (e *baseImageLoadEvent) Next(images ...ImageLoadEvent) ImageLoadEvent {
+	e.nextImages = images
```

Comment:
shouldn't this be an append? or fail if nextImages is already set? or should Next be called SetNext to fit more golang naming conventions?

Reply:
SetNext looked a bit strange when I was writing the tests for this, which is why it ended up as Next. Given the information that the API is intended to capture, I wonder whether this should be a verb of some kind instead, so it looks something like this:
```
    NewImageLoadEvent("/shimx64.efi").Loads(
        NewImageLoadEvent("/grubx64.efi").Loads(
            NewImageLoadEvent("/kernel1.efi"),
            NewImageLoadEvent("/kernel2.efi"),
        )
    )
```

Although in this case, `ImageLoadEvent` sounds a bit weird as well - maybe it should be `ImageLoadData` instead.

Reply:
yes, a verb would be clearer, not a fan of struct called *Data though, one option s/Event/Activity/

Reply:
I've called this `ImageLoadActivity` and changed the method to `Loads` now.